server/src/main/java/com/broadcom/lsp/cobol/core/visitor/VariableUsageDelegate.java

```diff
@@ -0,0 +1,171 @@
+import com.broadcom.lsp.cobol.core.model.Locality;
+import com.broadcom.lsp.cobol.core.model.SyntaxError;
+import com.broadcom.lsp.cobol.core.model.variables.Variable;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.RuleContext;
+import org.antlr.v4.runtime.Token;
+
+import java.util.*;
+
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * This class process the variable usage contexts. It accumulate usages when they appear in the text. Later it checks
+ * variable definition and updates its usages.
+ */
+@Slf4j
+@RequiredArgsConstructor
+class VariableUsageDelegate {
+  private List<VariableUsage> variableUsages = new ArrayList<>();
+  private final Map<Token, Locality> positionMapping;
+  private final MessageService messageService;
+
+  /**
+   * Accumulates variable appearance for analise its definition later.
+   *
+   * @param dataName the variable name
+   * @param locality the variable text position
+   * @param ctx the ANTLR variable context
+   */
+  void handleDataName(String dataName, Locality locality, CobolParser.QualifiedDataNameFormat1Context ctx) {
+    List<CobolParser.QualifiedInDataContext> hierarchy = ctx.qualifiedInData();
+    String qualifier = createQualifier(dataName, hierarchy);
+    Map<String, Token> parentVariables = collectParentVariables(hierarchy);
+    variableUsages.add(new VariableUsage(dataName, qualifier, locality, parentVariables));
+  }
+
+  /**
+   * Accumulates variable appearance for analise its definition later.
+   *
+   * @param dataName the variable name
+   * @param locality the variable text position
+   */
+  void handleTableCall(String dataName, Locality locality) {
+    variableUsages.add(new VariableUsage(dataName, ".* " + dataName, locality, Map.of()));
+  }
+
+  /**
+   * Finds correct definition for collected usages. Updates variable usage or produce an error.
+   *
+   * @param definedVariables the collection of COBOL variables
+   * @return the list of usage errors
+   */
+  List<SyntaxError> updateUsageAndGenerateErrors(Collection<Variable> definedVariables) {
+    List<SyntaxError> errors = new ArrayList<>();
+    for (VariableUsage variableUsage: variableUsages) {
+      List<Variable> foundVariables = findVariable(definedVariables, variableUsage.searchQualifier);
+      if (foundVariables.size() == 1) {
+        Variable variable = foundVariables.get(0);
+        variable.addUsage(variableUsage.locality);
+        addParentVariablesUsage(variableUsage.parentVariables, variable);
+      } else {
+        errors.add(createInvalidDefinition(variableUsage.name, variableUsage.locality));
+      }
+    }
+    return errors;
+  }
+
+  /**
+   * This value class used for temporal storage variable usages.
+   */
+  @Value
+  private static class VariableUsage {
```

Comment:
I think it would be better to keep the nested class declaration at the bottom of this class